### ISSUE TITLE
KeyError private.4 interface

### ISSUE THREAD

**Issue by anarchy9388** [anarchy9388/toroi#5]
@ikemen18  Like this
```
 python toroi.py tech_consult
Traceback (most recent call last):
  File "toroi.py", line 38, in <module>
    from libs.svc_base import Common, Common_Params
  File "/home/admin/toroi-1.5/libs/svc_base.py", line 25, in <module>
    from libs.parsers import DTParse
ImportError: No module named parsers
admin@g3ppss1042-n1-mgmt:~/toroi-1.5> sudo python toroi.py tech_consult
Traceback (most recent call last):
  File "toroi.py", line 1027, in <module>
    toroi_inst = Toroi(params)
  File "toroi.py", line 64, in __init__
    self.cluster_master_ip = self.get_cluster_master_ip()
  File "toroi.py", line 159, in get_cluster_master_ip
    cluster_master_ip = self.vdcs['self'].nodes[cluster_master].interfaces['private.4']
KeyError: 'private.4'
toroi-1.5> 
```

**Comment by anarchy9388** [comment]
```
Traceback (most recent call last):
  File "toroi.py", line 1027, in <module>
    toroi_inst = Toroi(params)
  File "toroi.py", line 64, in __init__
    self.cluster_master_ip = self.get_cluster_master_ip()
  File "toroi.py", line 159, in get_cluster_master_ip
    cluster_master_ip = self.vdcs['self'].nodes[cluster_master].interfaces['private.4']
KeyError: 'private.4'
```

**Comment by ikemen18** [comment]
This appears to be something related to rest API.  Getting the same error on a newer version of code.

**Comment by ikemen18** [comment]
Just confirmed it has to do with Rest API. If there is an error/WARNING flagged when checking, it fails out. The error on the appliance that threw this is now running toroi fine. No rest API errors are flagging.

**Comment by anarchy9388** [comment]
@ikemen18 

Here's what the structure looks like for the node object:
```
        "r120n4": {
            "capacity": "<undiscovered>",
            "deviceIds": "<undiscovered>",
            "disks": "<undiscovered>",
            "hostname": "lglf134.lss.emc.com",
            "id": "4fda3c08-0c97-46d0-b395-8a4bb5114203",
            "interfaces": {
                "private": "192.168.219.4",
                "private.4": "169.254.120.4",
                "private_ipmi": "192.168.219.104",
                "public": "10.246.151.134",
                "public.repl": "10.246.152.134",
                "remote_ipmi": "10.246.23.134"
```

I'm not sure why private.4 is missing in some cases, but we can workaround it for now.
Change the code for the get_cluster_master_ip to handle it:

```
    def get_cluster_master_ip(self):
        racknums = []
        for nodeKey, node in self.vdcs['self'].nodes.iteritems():
            racknums.append(node.racknum)
        racknums = list(set(racknums))
        cluster_master = 'r{}n1'.format(min(racknums))
        try:
            cluster_master_ip = self.vdcs['self'].nodes[cluster_master].interfaces['private.4']
        except KeyError:
            cluster_master_ip = self.vdcs['self'].nodes[cluster_master].interfaces['public']
        return cluster_master_ip
```

**Comment by anarchy9388** [comment]
@ikemen18 

Also this:
```
    def upgrade_service_console(self):
        upgrade_service_console = False
        self.msg_output('Checking service-console version and upgrading if version is not {}'.format(
            self.latest_service_console))
        self.msg_output('{}{}{}'.format(self.colors.BLUE, ('-' * 100), self.colors.ENDC))
-        cluster_master_ip = self.cluster_master_ip
        cluster_master = self.cluster_master
+        installer_node = self.localvdc.nodes[cluster_master]

        # Check Console Version
-        if self.vdcs['self'].nodes[cluster_master].versions.service_console != self.latest_service_console:
+        if installer_node.versions.service_console != self.latest_service_console:
            upgrade_service_console = True
        if DEBUG:
            upgrade_service_console = True
        # Upgrade Console
        if upgrade_service_console:
-            self.msg_output('Upgrading service-console on {}'.format(cluster_master_ip), indent=1)
+            self.msg_output('Upgrading service-console on {}'.format(), indent=1)
            result = self.copy_to_nodes(self.service_console_path,
                                        targetdir='/tmp/service_console/',
-                                        MACHINES=cluster_master_ip)
+                                        MACHINES=cluster_master)
            result = self.copy_to_nodes(self.service_console_path,
                                        targetdir='/home/admin',
-                                        MACHINES=cluster_master_ip)
+                                        MACHINES=cluster_master)
            self.msg_output('Copying service-console file..', indent=1, newline=False)
-            if result[cluster_master_ip]:
+            if result[cluster_master]:
                self.print_done()
            else:
                self.msg_output('Failed to copy service-console file', indent=1, error=True)
                exit(100)

            self.msg_output('Extracting service-console tgz..', indent=1, newline=False)
            cmd = 'tar -zxvf /home/admin/{} -C /home/admin'.format(self.service_console_path.split('/')[-1])
-            result = self.run_cmd(cmd, MACHINES=cluster_master_ip)
-            if result[cluster_master_ip]['retval'] == 0:
+            result = installer_node.run_cmd(command=cmd)
+            if result['retval'] == 0:
                self.print_done()
            else:
                self.msg_output('Failed to extract service-console tgz', indent=1, error=True)
                self.msg_output('{}'.format(result), indent=1, error=True)
                exit(101)

            self.msg_output('Installing new service-console..', indent=1, newline=False)
            cmd = '/home/admin/service-console install --force'
-            result = self.run_cmd(cmd, MACHINES=cluster_master_ip)
-            if 'Installation successfully finished' in result[cluster_master_ip]['stdout']:
+            result = installer_node.run_cmd(cmd)
+            if 'Installation successfully finished' in result['stdout']:
                self.print_done()
            else:
                self.msg_output('Failed to install new service-console:', indent=1, error=True)
                self.msg_output('{}'.format(result), indent=1, error=True)

            self.msg_output('service-console has successfully been updated.', indent=1)
            cmd = 'service-console -v'
-            result = self.run_cmd(cmd, MACHINES=cluster_master_ip)
-            self.msg_output('{}'.format(result[cluster_master_ip]['stdout']), indent=1)
+            result = installer_node.run_cmd(command=cmd)
+            self.msg_output('{}'.format(result['stdout']), indent=1)
        else:
            self.msg_output('service-console is up-to-date.', indent=1)
        self.msg_output()
```

**Comment by ikemen18** [comment]
Made the changes and tried to run with SC not installed: 

 `Traceback (most recent call last):
  File "toroi.py", line 1033, in <module>
    toroi_inst.main()
  File "toroi.py", line 993, in main
    self.do_pre_check()
  File "toroi.py", line 513, in do_pre_check
    self.upgrade_service_console()
  File "toroi.py", line 114, in upgrade_service_console
    self.msg_output('Upgrading service-console on {}'.format(), indent=1)
IndexError: tuple index out of range
`

**Comment by ikemen18** [comment]
Got it, needed to add cluster_master into the .format(), published to 1.6.
